l4: Fix l7rules ordering test failure

joestringer opened the pull request:
This patch fixes issue #1902 reported on GitHub where the ordering of
l7rules formatted to JSON is unreliable. We don't serialize the
L7DataMap to JSON very frequently, so solve this by sorting the
endpointselectors before assembling the JSON.

Fixes: #1902 
Reported-by: Ian Vernon <ian@cilium.io>
Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
l4: Fix l7rules ordering test failure

This patch fixes issue #1902 reported on GitHub where the ordering of
l7rules formatted to JSON is unreliable. We don't serialize the
L7DataMap to JSON very frequently, so solve this by sorting the
endpointselectors before assembling the JSON.

Reported-by: Ian Vernon <ian@cilium.io>
Signed-off-by: Joe Stringer <joe@covalent.io>

ianvernon reviewed (approved):
LGTM, thanks for taking care of this so quickly!

aanm commented on pkg/policy/l4.go:
```diff
@@ -39,13 +40,25 @@ func (l7 L7DataMap) MarshalJSON() ([]byte, error) {
-		buffer.WriteString(k.LabelSelectorString())
+		buffer.WriteString(es.LabelSelectorString())
 		buffer.WriteString("\": ")
-		b, err := json.MarshalIndent(v, "    ", "  ")
```
This line is unnecessary, as well all the `\n` and spaces since this function is `MarshalJSON`, not `MarshalIndent`

https://golang.org/pkg/encoding/json/#MarshalIndent

joestringer replied:
Fixed.

joestringer committed:
l4: Fix L7DataMap.MarshalJSON() whitespacing.

L7DataMap.MarshalJSON() has been attempting to implement
MarshalIndent() rather than MarshalJSON(), but this is unnecessary -
json.MarshalIndent() will do this for us. Remove the whitespace
formatting from this function.

Signed-off-by: Joe Stringer <joe@covalent.io>

rlenglet commented on pkg/policy/l4.go:
```diff
@@ -39,24 +40,36 @@ func (l7 L7DataMap) MarshalJSON() ([]byte, error) {
 
+	/* First, create a sorted slice of the selectors so we can get
+	 * consistent JSON output */
+	selectors := make(api.EndpointSelectorSlice, len(l7))
```
Here you create a slice of *length* `len(l7)`. You should create an empty slice of *capacity* `len(l7)`:
```
selectors := make(api.EndpointSelectorSlice, 0, len(l7))
```

I'm surprised that the tests are passing...

rlenglet commented on pkg/policy/l4.go:
```diff
@@ -39,24 +40,36 @@ func (l7 L7DataMap) MarshalJSON() ([]byte, error) {
-		buffer.WriteString("\": ")
-		b, err := json.MarshalIndent(v, "    ", "  ")
+	for _, es := range selectors {
+		if _, ok := l7[es]; !ok {
```
Alright, this is why the tests are passing, you're ignoring the `len(l7)` `nil` selectors that are always at the beginning of your slice. You should be able to remove this after you fix the slice creation above?

joestringer replied:
Nice catch, yup! I'll fix this up.
